Lib/email/feedparser.py

```diff
@@ -87,40 +87,261 @@ def readline(self):
+
+        # We're dumping. There should be absolutely no other pending lines,
+        # because those should've been dumped.
+        assert not self._lines
+        if not self._eofstack:
+            # There's nothing that will ever tell us to stop dumping. Dump away
+            return True
+
+        all_boundary_matches = True
+        for pred in self._eofstack:
+            if not hasattr(pred, 'is_boundary_match'):
+                return False
+
+        # We only care about boundaries; we can dump as long as there's no
+        # potential boundaries.
+        return not line.startswith("-", start, end)
+
+    def _is_dump_midline(self):
+        if not self._dump_destination:
+            return False
+
+        return self._dump_destination[-1][-1] not in ('\n', '\r')
+
+    def _push_data(self, data):
+        # Find first newline character in the data
+        unl_start_index = BufferedSubFile._find_unl(data)
+        if unl_start_index < 0:
             # No new complete lines, wait for more.
+            # Check to see if we had a previous dangling partial newline
+            if self._dangling_partial:
+                # We previously pushed a dangling line expecting \n to follow,
+                # however we received other data instead. Therefore, that \r
+                # does actually terminate a line. Go ahead and push it.
+                self._flush_partial()
+
+            # No lines in data to push; wait for more data
+            if self._is_dump_midline():
+                assert not self._partial
+                self._dump_destination.append(data)
+            else:
+                self._partial.append(data)
             return
 
-        # Crack into lines, preserving the linesep characters.
-        self._partial.seek(0)
-        parts = self._partial.readlines()
-        self._partial.seek(0)
-        self._partial.truncate()
+        data_start_index = 0
+
+        # Complete our previous/partial line
+        if self._partial:
+            if self._dangling_partial:
+                if data[0] != NL:
+                    # "\r<whatever>" -- push what we had, it's been terminated
+                    self._flush_partial()
+                else:
+                    # "\r\n" -- append \n to complete it and push
+                    self._partial.append(NL)
+                    self._flush_partial()
+                    data_start_index = 1
+
+                    # Find the next newline
+                    unl_start_index = BufferedSubFile._find_unl(
+                        data, data_start_index)
+            else:
+                # Complete our partial with the new line and push it
+                unl_end_index = BufferedSubFile._find_unl_end(
+                    data, unl_start_index)
+                if unl_end_index < 0:
+                    # The newline is incomplete; append data and return
+                    self._partial.append(data)
+                    self._dangling_partial = True
```

Comment:
I would suggest having a method doing the partial.append() and the dangling_partial=True (AFAICT, doing the append should always set the flag)

Reply:
This would make sense; the thought also later occurred to me that we can just have a property which checks "is there a partial, and does it end in `'\r'`?", but I didn't want to re-benchmark this at the time to see if there was any measurable difference. I'm a little less worried now though, and tracking less explicit state may be more maintainable. It is admittedly error-prone.

We only set the flag though when the line is ending on a `'\r'` specifically, not on every append(). Here, we know the last character is `'\r'` because the universal newline is present, but lacks an end (so we need to see if a `'\n'` follows).